src/plugins/expressions/common/execution/execution.ts

```diff
@@ -202,147 +205,143 @@ export class Execution<
   public start(input: Input = null as any) {
     if (this.hasStarted) throw new Error('Execution already started.');
     this.hasStarted = true;
-
     this.input = input;
     this.state.transitions.start();
 
-    const { resolve, reject } = this.firstResultFuture;
-    const chainPromise = this.invokeChain(this.state.get().ast.chain, input);
+    this.firstResultFuture = this.race(this.invokeChain(this.state.get().ast.chain, input)).pipe(
+      catchError((error) => {
+        if (this.abortController.signal.aborted) {
+          this.childExecutions.forEach((execution) => execution.cancel());
 
-    this.race(chainPromise).then(resolve, (error) => {
-      if (this.abortController.signal.aborted) {
-        this.childExecutions.forEach((ex) => ex.cancel());
-        resolve(createAbortErrorValue());
-      } else reject(error);
-    });
+          return of(createAbortErrorValue());
+        }
 
-    this.firstResultFuture.promise
-      .then(
-        (result) => {
-          if (this.context.inspectorAdapters.expression) {
-            this.context.inspectorAdapters.expression.logAST(this.state.get().ast);
-          }
+        return throwError(error);
+      }),
+      tap({
+        next: (result) => {
+          this.context.inspectorAdapters.expression?.logAST(this.state.get().ast);
           this.state.transitions.setResult(result);
         },
-        (error) => {
-          this.state.transitions.setError(error);
-        }
-      )
-      .finally(() => {
-        this.abortRejection.cleanup();
-      });
-  }
-
-  async invokeChain(chainArr: ExpressionAstFunction[], input: unknown): Promise<any> {
-    if (!chainArr.length) return input;
-
-    for (const link of chainArr) {
-      const { function: fnName, arguments: fnArgs } = link;
-      const fn = getByAlias(this.state.get().functions, fnName);
-
-      if (!fn) {
-        return createError({
-          name: 'fn not found',
-          message: i18n.translate('expressions.execution.functionNotFound', {
-            defaultMessage: `Function {fnName} could not be found.`,
-            values: {
-              fnName,
-            },
-          }),
-        });
-      }
+        error: (error) => this.state.transitions.setError(error),
+      }),
+      finalize(() => this.abortRejection.cleanup()),
+      share()
```

Comment:
Wondering if we might want to use `shareReplay`? 🤔

Reply:
It makes total sense. I fixed that. Thanks!